crates/gitbutler-repo/src/repository_ext.rs

```diff
@@ -197,11 +196,28 @@ impl RepositoryExt for git2::Repository {
             } else if (status.is_index_deleted() && !status.is_wt_new()) || status.is_wt_deleted() {
                 tree_update_builder.remove(path);
             } else {
-                let file_path = worktree_path.join(path);
-                let file = std::fs::read(file_path)?;
-                let blob = self.blob(&file)?;
+                let file_path = worktree_path.join(path).to_owned();
 
-                tree_update_builder.upsert(path, blob, git2::FileMode::Blob);
+                if file_path.is_symlink() {
+                    let resolved_path = file_path.read_link()?;
+                    let path_str = resolved_path
+                        .to_str()
+                        .context("Failed to convert path to str")?;
+
+                    let blob = self.blob(path_str.as_bytes())?;
+                    tree_update_builder.upsert(path, blob, git2::FileMode::Link);
+                } else {
+                    let file = std::fs::read(&file_path)?;
+                    let blob = self.blob(&file)?;
+
+                    let file_type = if is_executable(&file_path.metadata()?) {
```

Comment:
nit: I'd get `file_path.metadata()` first, and do the file-kind checks through that. `file_path.is_symlink()` fetches its own under the hood and I don't think it's cached. Right now, this means two `stat` calls for non-symlinks, instead of one. It's a nit because this isn't a hot loop.